src/validators/decimal.rs

```diff
@@ -0,0 +1,254 @@
+use crate::errors::ValError;
+use crate::errors::ValResult;
+use crate::input::Input;
+use crate::recursion_guard::RecursionGuard;
+use crate::tools::SchemaDict;
+use crate::PydanticCustomError;
+
+use super::{BuildValidator, CombinedValidator, Definitions, DefinitionsBuilder, Extra, Validator};
+
+#[derive(Debug, Clone)]
+pub struct DecimalValidator {
+    strict: bool,
+    allow_inf_nan: bool,
+    check_digits: bool,
+    multiple_of: Option<Py<PyAny>>,
+    le: Option<Py<PyAny>>,
+    lt: Option<Py<PyAny>>,
+    ge: Option<Py<PyAny>>,
+    gt: Option<Py<PyAny>>,
+    max_digits: Option<u64>,
+    decimal_places: Option<u64>,
+    decimal_type: Py<PyType>,
+}
+
+impl BuildValidator for DecimalValidator {
+    const EXPECTED_TYPE: &'static str = "decimal";
+    fn build(
+        schema: &PyDict,
+        config: Option<&PyDict>,
+        _definitions: &mut DefinitionsBuilder<CombinedValidator>,
+    ) -> PyResult<CombinedValidator> {
+        let py = schema.py();
+        let allow_inf_nan = schema_or_config_same(schema, config, intern!(py, "allow_inf_nan"))?.unwrap_or(false);
+        let decimal_places = schema.get_as(intern!(py, "decimal_places"))?;
+        let max_digits = schema.get_as(intern!(py, "max_digits"))?;
+        if allow_inf_nan && (decimal_places.is_some() || max_digits.is_some()) {
+            return Err(PyValueError::new_err(
+                "allow_inf_nan=True cannot be used with max_digits or decimal_places",
+            ));
+        }
+        Ok(Self {
+            strict: is_strict(schema, config)?,
+            allow_inf_nan,
+            check_digits: decimal_places.is_some() || max_digits.is_some(),
+            decimal_places,
+            multiple_of: schema.get_as(intern!(py, "multiple_of"))?,
+            le: schema.get_as(intern!(py, "le"))?,
+            lt: schema.get_as(intern!(py, "lt"))?,
+            ge: schema.get_as(intern!(py, "ge"))?,
+            gt: schema.get_as(intern!(py, "gt"))?,
+            max_digits,
+            decimal_type: py
+                .import(intern!(py, "decimal"))?
+                .getattr(intern!(py, "Decimal"))?
+                .extract()?,
+        }
+        .into())
+    }
+}
+
+impl_py_gc_traverse!(DecimalValidator {
+    multiple_of,
+    le,
+    lt,
+    ge,
+    gt,
+    decimal_type
+});
+
+impl Validator for DecimalValidator {
+    fn validate<'s, 'data>(
+        &'s self,
+        py: Python<'data>,
+        input: &'data impl Input<'data>,
+        extra: &Extra,
+        _definitions: &'data Definitions<CombinedValidator>,
+        _recursion_guard: &'s mut RecursionGuard,
+    ) -> ValResult<'data, PyObject> {
+        // FIXME might need to check input type rather than blindly call, esp in strict mode
+        let decimal = input.validate_decimal(extra.strict.unwrap_or(self.strict), &self.decimal_type)?;
+        // FIXME use PyO3 raw ffi PyObject_CallOneArg ?
+        // let decimal = self.decimal_type.as_ref(py).call1((either_decimal,))?;
```

Comment:
i guess this comment can be removed, if such a call is required, it should be within `validate_decimal` in the python implementation.